diff -Nur httpd-2.4.48-orig/server/core.c httpd-2.4.48/server/core.c
--- httpd-2.4.48-orig/server/core.c	2021-05-12 11:14:42.000000000 +0100
+++ httpd-2.4.48/server/core.c	2021-06-02 17:38:25.539552064 +0100
@@ -1754,7 +1754,7 @@
             opt = OPT_INCLUDES;
         }
         else if (!ap_cstr_casecmp(w, "FollowSymLinks")) {
-            opt = OPT_SYM_LINKS;
+            opt = OPT_SYM_OWNER;
         }
         else if (!ap_cstr_casecmp(w, "SymLinksIfOwnerMatch")) {
             opt = OPT_SYM_OWNER;
@@ -2003,7 +2003,7 @@
             opt = OPT_INCLUDES;
         }
         else if (!ap_cstr_casecmp(w, "FollowSymLinks")) {
-            opt = OPT_SYM_LINKS;
+            opt = OPT_SYM_OWNER;
         }
         else if (!ap_cstr_casecmp(w, "SymLinksIfOwnerMatch")) {
             opt = OPT_SYM_OWNER;
